### PR TITLE
(2835) Feature flag the activity linking for the import CSVs

### PR DESCRIPTION
**Pull request by jkempster34**
### Decouple Import class from `activity_csv_columns.csv`
We want to be able to feature flag one of the CSV columns (activity linking).
However, this is difficult currently. Including the feature flag in the
`activity_csv_columns` initializer means that the effects of the feature flag
won't apply until the app has been restarted. Including the feature flag in the
current large Import class will be messy.

This decouples the rest of the `Import` class from the
`activity_csv_columns.csv` file, and provides an interface for common
operations.

Note: the `invalid_for_level_b_ispf_non_oda` and
`invalid_for_level_c_d_ispf_non_oda` are left as they were for now, but are
candidates for further refactoring.

### Feature flag the activity linking feature in the import CSVs
We have put the activity linking feature behind a feature flag in the UI. We
have had some feedback that linking is confusing to POs while
there's no ODA funds being spent.

This does the same for the import CSVs.

## Next steps

- [ ] Is an ADR required? An ADR should be added if this PR introduces a change to the architecture.
- [ ] Is a changelog entry required? An entry should always be made in `CHANGELOG.md`, unless this PR is a small tweak which has no impact outside the development team.
- [ ] Do any environment variables need amending or adding?
- [ ] Have any changes to the XML been checked with the IATI validator? See [XML Validation](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/develop/doc/xml-validation.md)

### CODE REVIEW

File: app/models/activity/import/field.rb
Comment (jkempster34) [inline app/models/activity/import/field.rb:1]: How do we feel about the interface for this class? I feel I got a bit carried away

**Comment by jkempster34** [comment]
[Feature flag the linking feature in the various CSVs](https://trello.com/c/cCSmtFsP/2835-feature-flag-the-linking-feature-in-the-various-csvs)

**Review by jkempster34 — COMMENTED** [review]
Uploading a CSV with a linked_activity_id when the feature is turned off will silently fail, i.e., the activity will not be linked, and there will be no error message.

I reckon that's fine?

File: spec/models/activity/import/field_spec.rb
Comment (yndajas) [inline spec/models/activity/import/field_spec.rb:10]: `be_empty` or `be_nil`?

File: spec/models/activity/import/field_spec.rb
Comment (yndajas) [inline spec/models/activity/import/field_spec.rb:97]: Hmm I can see how `expected_fields` is useful in this case

**Review by yndajas — COMMENTED** [review]
This looks good! I've made some comments/suggestions on the details, but in general it seems like a useful improvement in how we handle the import fields

Might be good to get some more eyes on it

**Comment by yndajas** [comment]
> Uploading a CSV with a linked_activity_id when the feature is turned off will silently fail, i.e., the activity will not be linked, and there will be no error message.
> 
> I reckon that's fine?

They would have to manually add the `linked_activity_id` in to the template to do so, right?

File: app/models/activity/import/field.rb
Comment (jkempster34) [reply app/models/activity/import/field.rb]: Thanks! I thought they were in `models`

File: spec/models/activity/import/field_spec.rb
Comment (jkempster34) [reply spec/models/activity/import/field_spec.rb]: It's not unfortunately!

File: spec/models/activity/import/field_spec.rb
Comment (jkempster34) [reply spec/models/activity/import/field_spec.rb]: Agreed, I've moved it into a `let(:field_headings) { subject.map(&:heading) }` which may still be contentious, but less so

File: spec/models/activity/import/field_spec.rb
Comment (jkempster34) [reply spec/models/activity/import/field_spec.rb]: You have mentioned below, but it's so I can add to it for the linked activity specs!

File: spec/models/activity/import/field_spec.rb
Comment (jkempster34) [reply spec/models/activity/import/field_spec.rb]: Select returns an empty array when not found, there may be another tidier way of doing this though

File: app/models/activity/import/field.rb
Comment (jkempster34) [reply app/models/activity/import/field.rb]: Thanks for the review! I've made most of the changes.

So, part of this is me projecting my desire for it to look like ActiveRecord, but because they are keyword arguments, they have to be included, so really the signature is `find_by(attribute_name:)`.

However, I totally get your point, and I did wonder what would happen when somebody wanted to find by heading because `find_by(heading:)` wouldn't be allowed.

**Comment by jkempster34** [comment]
> > Uploading a CSV with a linked_activity_id when the feature is turned off will silently fail, i.e., the activity will not be linked, and there will be no error message.
> > I reckon that's fine?
> 
> They would have to manually add the `linked_activity_id` into the template to do so, right?

I was thinking if they were using a template they had downloaded previously. Not sure if it's a massive problem though

File: app/models/activity/import/field.rb
Comment (Gweaton) [reply app/models/activity/import/field.rb]: I like this too - that `Activity::Import` class and its tests have been getting more and more unwieldy, so breaking them down into smaller chunks is much appreciated, and this feels like a good place to split out some of this functionality.

File: spec/services/activity/import/field_spec.rb
Comment (Gweaton) [inline spec/services/activity/import/field_spec.rb:7]: I have _opinions_ on the readability of `described_class` in general, but for this type-checking tests i.e. `to be_a(Field)` I prefer explicitly asserting on the name of the class as I find it more readable in the context of an individual test. Is there something `described_class` gives us over the use of the class name specifically, or is more of a convention?

**Review by Gweaton — APPROVED** [review]
This is a really nice refactor, thanks for your hard work. I'll approve but might be worth waiting for Ynda's approval before we merge.

File: spec/services/activity/import/field_spec.rb
Comment (jkempster34) [reply spec/services/activity/import/field_spec.rb:7]: I used to work on a project with a very strict Rubocop, so I've picked up some habits from that. No particular benefit, happy to change it.

File: app/models/activity/import/field.rb
Comment (yndajas) [reply app/models/activity/import/field.rb]: Good point re: signature. I guess I typically try to make the method name descriptive enough in itself, but you're right that the keyword arguments are part of the signature. But as it stands, couldn't you switch the method names of `find_all_by` and `where`? That feels a little strange to me

File: spec/models/activity/import/field_spec.rb
Comment (yndajas) [reply spec/models/activity/import/field_spec.rb]: Ah true. Given you're just looking for one field, would be make more sense to use `.find` here and then assert on `nil` (in both tests)?

**Comment by yndajas** [comment]
> I was thinking if they were using a template they had downloaded previously. Not sure if it's a massive problem though

I think since ISPF isn't live in RODA yet, it should be fine (to ignore)

**Pull request merged**
